Tests/tests/test_FrontPanel.cpp

```diff
@@ -0,0 +1,514 @@
+    }
+
+    virtual ~FrontPanelInitializedEventTest() override
+    {
+        dispatcher->Deactivate();
+        dispatcher->Release();
+
+        PluginHost::IFactories::Assign(nullptr);
+    }
+};
+
+class FrontPanelInitializedEventDsTest : public FrontPanelInitializedEventTest {
+protected:
+    FrontPanelIndicatorImplMock frontPanelIndicatorImplStringMock;
+    FrontPanelIndicatorImplMock frontPanelIndicatorImplIntMock;
+    FrontPanelTextDisplayImplMock frontPanelTextDisplayImplStringMock;
+    FrontPanelTextDisplayImplMock frontPanelTextDisplayImplMock;
+    FrontPanelTextDisplayImplMock frontPanelTextDisplayImplIntMock;
+
+
+    FrontPanelInitializedEventDsTest()
+        : FrontPanelInitializedEventTest()
+    {
+        device::FrontPanelIndicator::getInstance("Power").impl = &frontPanelIndicatorImplStringMock;
+        device::FrontPanelIndicator::getInstance(0).impl = &frontPanelIndicatorImplIntMock;
+        device::FrontPanelTextDisplay::getInstance("Text").impl = &frontPanelTextDisplayImplStringMock;
+        device::FrontPanelTextDisplay::getInstance(0).impl = &frontPanelTextDisplayImplIntMock;
+        device::FrontPanelTextDisplay::getInstance().impl = &frontPanelTextDisplayImplMock;
+	
+    }
+
+    virtual ~FrontPanelInitializedEventDsTest() override
+    {
+        device::FrontPanelIndicator::getInstance("Power").impl = nullptr;
+        device::FrontPanelTextDisplay::getInstance("Text").impl = nullptr;
+        device::FrontPanelTextDisplay::getInstance().impl = nullptr;
+	device::FrontPanelTextDisplay::getInstance(0).impl = nullptr;
+        device::FrontPanelIndicator::getInstance(0).impl = nullptr;
+
+    }
+};
+
+TEST_F(FrontPanelTest, RegisteredMethods)
+{
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setBrightness")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("getBrightness")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("powerLedOn")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("powerLedOff")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setClockBrightness")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("getClockBrightness")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("getFrontPanelLights")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("getPreferences")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setPreferences")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setLED")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setBlink")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("set24HourClock")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("is24HourClock")));
+    EXPECT_EQ(Core::ERROR_NONE, handler.Exists(_T("setClockTestPattern")));
+}
+
+TEST_F(FrontPanelDsTest, setBrightnessWIndex)
+{
+    EXPECT_EQ(Core::ERROR_NONE, handler.Invoke(connection, _T("setBrightness"), _T("{\"brightness\": 1,\"index\": \"power_led\"}"), response));
+    EXPECT_EQ(response, string("{\"success\":true}"));
```

Comment:
a test case should actually test something. how do you verify the business logic of "setBrightness"? i.e. how do you verify it works correctly instead of returning `"{\"success\":true}"`. by looking at this test case I assume "setBrightness" may just return `"{\"success\":true}"` and that will be OK for your test suite? if you are verifying that "setBrightness" triggers the backend then please add expectations for that. this concerns other test cases as well

Reply:
In this case, with setBrightness(and most of the other functions that look like this), the function, depending on the inputs, calls something akin to this(in either FrontPanel.cpp or frontpanel.cpp):

`try
                            {
                                device::FrontPanelIndicator::getInstance(fp_ind.c_str()).setBrightness(int(brightness));
                                ok = true;
                            }`
   where it calls our mocked function. The set functions never expect any returned input, since our mock functions don't 'fail' so to speak, the only things to test are the different inputs(whether it has index, and what value index contains changes what parts of the function we go in to) that are sent to the function, so we trigger the different parts of the code, and make sure, assuming the mock is successful, the function still runs through everything appropriately. As the set functions are all pretty much type void, they can't return anything, so I'm not sure what you want me to look for in this case. 
   
   Looking at other test classes, test_HdmiInput.cpp, test_FrameRate.cpp, and a few others have functions that just pass in inputs, and expect no errors and `"{\"success\":true}"`. So I'm not sure what else there is to verify as the functions in question never do anything with the output from setBrightness, and simply call it--and if it succeeds, sets it so that the function returns true.

Reply:
OK this test case verifies JSON RPC interface. got it

on set-top one can check led brightness to see that "setBrightness" works. if `device::FrontPanelIndicator` switches led in real life, you can add another test case with `ON_CALL/EXPECT_CALL` to verify that "setBrightness" does some job on that matter as well

Reply:
Yes, the JSON RPC interface. Do you have an example of what you want for the other test case, but I'm still not sure what I should be verifying here.

Also, in regards to the Valgrind reports, I'm assuming you mean the loss record where some of the bytes are referred to as 'still reachable'? I remember looking into them before, but I couldn't figure out what was causing them to remain still reachable. Backtracing, they all seem to be objects created from things defined in frontpanel.h, and not something I'm directly calling/creating on my end.

Reply:
@hgfell683 

setBrightnessWIndex
setBrightnessClock
setBrightness
setBrightnessFalse
setBrightnessNeg1

that you added are verifying that "setBrightness" checks input parameters

checking input parameters is not the only thing "setBrightness" does

it also should have some code that sets the brightness

please add a test for the code which sets brightness. so that we know that "setBrightness" sets brightness. this is what this request is for

'still reachable' may not be leaked but can indicate something wrong in the code. should be investigated, what causes that. this is what the second request is for

Reply:
@npoltorapavlo the setBrightness test functions still tests set the brightness. The value that always needs to be sent, regardless, is brightness (in most cases, it's brightness = 1, when it isn't set it errors out, and there is a case to test specifically for that), and that value is what is being used to set the brightness, and is passed into the setBrightness function that we mocked in devicesettings.h. What the index is really telling the code, by which section it goes into, is what light we're talking about. Whether we're setting the brightness for power led, the clock led, or another led that is present on the device. But regardless of which section it goes int based on the index, we always set trigger the setBrightness mock function. If it's not clear, I can rewrite the the names so that it's a little more clear on what this does. And if I'm still misunderstanding what you're asking, I apologize in advance.

And I'll take another look and see if I can figure out what is causing the memory issues. Fresh eyes from a relaxing weekend might help me see what I couldn't see last week.

Reply:
something like this would work:
```
TEST_F(FrontPanelDsTest, setBrightnessTriggersBrightnessViaDs)
{
    EXPECT_CALL(frontPanelIndicator...Mock, setBrightness(...))
        .Times(1)
        .WillOnce(::testing::Return());

    EXPECT_EQ(Core::ERROR_NONE, handler.Invoke(connection, _T("setBrightness"), ...
    EXPECT_EQ(response, string("{\"success\":true}"));
}
```